### DISCUSSION TITLE
CMake - ImGUI::ImageButton/Image linking

### DISCUSSION THREAD

**Discussion by edvn0** [ocornut/imgui#5052] [New Users Build/Link/Run issues ONLY!]
Hi. It might be the case that I'm doing something terribly obvious incorrect, but up until an accidental git-reset everything was linking fine and well.

My engine builds and links imgui via CMake, using add_subdirectory(vendor/imgui) (also using imguizmo, box2d, glfw and other non-imgui-related libraries). My applications link my engine via a simple target_link_library(my_app PRIVATE my_engine).

The problem is that whenever I change the linking of libraries into my engine from PUBLIC to either interface or private, my applications can no longer access ImGUI::ImageButton / Image.

For anyone well more involved with CMake than me, how do I mitigate this? Is PUBLIC okay to use?

**Comment by mathisloge** [comment]
I'm really recommending reading the documentation: https://cmake.org/cmake/help/latest/command/target_link_libraries.html#libraries-for-a-target-and-or-its-dependents

you would use private if all headers etc. are not part of the public consumable functions. so if your app depends on calling the dependencies of myengine, you would want `PUBLIC` since those dependencies should be added to the link interface.

**Reply by edvn0** [reply]
I see. I guess the question then becomes - does public linking of libraries incur any cost, be it size or speed, to the consumer? Is it preferable to have the consumer link against imgui "too", if you understand?

These may be stupid questions, but this is quite new to me.

**Reply by mathisloge** [reply]
If your libraries public interfaces have a public visibility of dependencies, you need to link public. If you dont want to add dependencies to the public link interfaces, you need to hide those in your private implementation.

If you have private link interfaces and those are public visible because your public interfaces depend on them, your consumers wont be able to use such functionalities except they link manually to those.

Gernerally you should only add libraries to your link interfaces which you use.

Any not used libraries will usually be stripped out from the linker.
